rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -183,27 +197,17 @@ void Player::play()
   try {
     do {
       reader_->open(storage_options_, {"", rmw_get_serialization_format()});
-      if (reader_->has_next()) {
-        // Reader does not have "peek", so we must "pop" the first message to see its timestamp
-        auto message = reader_->read_next();
-        prepare_clock(message->time_stamp);
-        // Make sure that first message gets played by putting it into the play queue
-        message_queue_.enqueue(message);
-      } else {
-        // The bag contains no messages - there is nothing to play
-        return;
-      }
-
-      // TODO(karsten1987): Put this work into the constructor
-      topic_qos_profile_overrides_ = play_options_.topic_qos_profile_overrides;
-      prepare_publishers();
+      const auto starting_time = std::chrono::duration_cast<std::chrono::nanoseconds>(
+        reader_->get_metadata().starting_time.time_since_epoch()).count();
 
       storage_loading_future_ = std::async(
         std::launch::async,
         [this]() {load_storage_content();});
 
       wait_for_filled_queue();
 
+      clock_->jump(starting_time);
+      clock_->resume();
```

Comment:
I don't think this makes sense to me - `play` implies a `resume` to me. Messages should always start when you call `play`, right? The scenario you are describing, to me, sounds like

* Create the Player, but don't call `play` - it's "paused"
* call `play` when you want it to start

Reply:
WRT. 
> Messages should always start when you call play, right? 

Nop. I think it could be a situation when someone want to have deferred play. Calling `play(..)`  will prepare everything to play and resume will trigger the real playback immediately. Inside `play(..)` method a lot of initialization routines and even spawning new tread and wait for storage to load content in the queue. \
At least it will be consistent according to the internal state of the player.

There are some situations when it's necessary to minimize timeout when you start playing data after some event happened. \
Consider the case if you have a huge bag file with sequential reader and want to start playing at the timestamp close to the end. It could take some substantial time when reader will adjust his position and preload data in buffers.

Reply:
Ok. Let's keep it as is for now. Otherwise we need to init clock with resume state on constructor.

Reply:
I actually recall a real use-case wen we will need playback starting in pause. \
This is "deterministic replay". It's supposed to be used in junction with `play_next()` message.

The idea is that for replay messages deterministically we need to control at what time to play next message. And this control is driven by the receiver side. We need to play next message only when we will get confirmation that all recipients have received previous on.  \
The flow would be: 
1. Create player
2. Put player in pause
3. Call Player::play(..) - player shall not call resume - otherwise it will start playing and we loose determinism in replay.
4. Call play_next() by event i.e. callback from services interface

I would suggest to init clock with `pause` equal to `false` in constructor and remove `resume()` in `player::play(..)`. In this case default behavior will be the same as in current implementation. The normal flow will be:
1. Create player
3. Call Player::play(..) - player shall not call resume. Playback will start automatically because settled `pause == false` in constructor

At the same time we will have opportunity for deterministic replay scenario with `play_next()` calling `Player::pause()` right before `Player::play()`.

Reply:
Ok - that makes sense to me, I will change this

Reply:
The latest revision should reflect this desired behavior.